gnovm/pkg/gnolang/values.go

```diff
@@ -2533,7 +2533,7 @@ func fillValueTV(store Store, tv *TypedValue) *TypedValue {
 			cv.Base = base
 			switch cb := base.(type) {
 			case *ArrayValue:
-				et := baseOf(tv.T).(*ArrayType).Elt
+				et := baseOf(tv.T).Elem()
```

Comment:
baseOf() is only about getting the base type of a (named) DeclaredType.
Since tv.V is a PointerValue, it means baseOf(tv.T) must be a PointerType.
(tv.T may not be a PointerType directly, but could also be a named declared type).
So .(*ArrayType) is in error, it should say baseOf(tv.T).(*PointerType).Elem().
The type conversion makes the line faster, as it doesn't need to run a generic interface.
.(*PointerType).Elt is probably as fast as .(*PointerType).Elem() due to inlining of .Elem().

Reply:
The high level gist of this case clause is as follows:
Normally we don't see this case because we fetch values directly from the Value's type and Value,
but only when the Value is an *ArrayType do we need to pass in the element type, 
(that is, ArrayValue.GetPointerAtIndexInt2() takes the element type as last argument),
because sometimes the ArrayValue is full of nil values and the type cannot be derived,
and we don't want to store the element type separately as an extra ArrayValue field
(because it is already available as you can see here, in the tv.T somehow, even if tv.T is a PointerType,
you can still get baseOf(tv.T).(*PointerType).Elt).

This is all made additionally confusing by the overloading of the term "base".
baseOf() is only about getting the underlying type for a DeclaredType, 
while PointerValues also have a "base" which is the base object within which the pointer points to.

The solution to this overloading is to rename baseOf() and similar references to base
(when referring to the underlying type of a declared type) to underlyingOf(),
while keeping PointerValue.base as base.

Reply:
Okay, thanks; that makes sense.  I've made the suggested change and will create a new issue for all of the renaming of `baseOf()` -> `underlyingOf` or maybe `underlyingTypeOf`?

While looking at this with @piux2 to we noticed that it seems to be the case where an entire array will be persisted if there is a reference to any of its elements persisted to the realm.  Is this desireable?  It only appears to be necessary to persist the element value in storage in the case where the array container itself is only transitory.  I made a rather naive approach to change this on [this branch](https://github.com/gnolang/gno/commit/67fa08ffedb5c0b81b692f47300d907472bfe1a4).  What do you think?

Reply:
> While looking at this with @piux2 to we noticed that it seems to be the case where an entire array will be persisted if there is a reference to any of its elements persisted to the realm. Is this desireable? It only appears to be necessary to persist the element value in storage in the case where the array container itself is only transitory. I made a rather naive approach to change this on [this branch](https://github.com/gnolang/gno/commit/67fa08ffedb5c0b81b692f47300d907472bfe1a4). What do you think?

here is the case 

https://github.com/gnolang/gno/issues/1675